### ISSUE TITLE
`Pack/unpack` fusion blocked by `collapse_shape/expand_shape`

### ISSUE THREAD

**Issue by pzread** [iree-org/iree#16682]
We found there are some un-fused pack/unpack ops in models and I think it is due to the `collapse_shape` and `expand_shape` between their fusable producer and consumer ops. For example:

```mlir
func.func @collapse_pack_and_unpack_expand(%arg0: tensor<4x3072x64xf32>, %arg1: tensor<3072xf32>, %arg2: tensor<32x256x8x1xf32>) -> tensor<256x12xf32> {
  %cst = arith.constant 0.000000e+00 : f32
  %0 = tensor.empty() : tensor<3072x64x4xf32>
  %1 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0, d1)>, affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%arg0, %arg1 : tensor<4x3072x64xf32>, tensor<3072xf32>) outs(%0 : tensor<3072x64x4xf32>) {
  ^bb0(%in: f32, %in_0: f32, %out: f32):
    %10 = arith.addf %in, %in_0 : f32
    linalg.yield %10 : f32
  } -> tensor<3072x64x4xf32>
  // Collapse before packing
  %collapsed = tensor.collapse_shape %1 [[0], [1, 2]] : tensor<3072x64x4xf32> into tensor<3072x256xf32>
  %2 = tensor.empty() : tensor<384x256x8x1xf32>
  %pack = tensor.pack %collapsed outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %2 : tensor<3072x256xf32> -> tensor<384x256x8x1xf32>
  %3 = tensor.empty() : tensor<384x32x8x8xf32>
  %4 = linalg.fill ins(%cst : f32) outs(%3 : tensor<384x32x8x8xf32>) -> tensor<384x32x8x8xf32>
  %5 = linalg.mmt4d ins(%pack, %arg2 : tensor<384x256x8x1xf32>, tensor<32x256x8x1xf32>) outs(%4 : tensor<384x32x8x8xf32>) -> tensor<384x32x8x8xf32>
  %6 = tensor.empty() : tensor<3072x256xf32>
  %unpack = tensor.unpack %5 outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 8] into %6 : tensor<384x32x8x8xf32> -> tensor<3072x256xf32>
  // Expand after unpacking
  %expanded = tensor.expand_shape %unpack [[0, 1], [2]] : tensor<3072x256xf32> into tensor<12x256x256xf32>
  %7 = tensor.empty() : tensor<256x12xf32>
  %8 = linalg.fill ins(%cst : f32) outs(%7 : tensor<256x12xf32>) -> tensor<256x12xf32>
  %9 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%expanded : tensor<12x256x256xf32>) outs(%8 : tensor<256x12xf32>) {
  ^bb0(%in: f32, %out: f32):
    %10 = arith.addf %out, %in : f32
    linalg.yield %10 : f32
  } -> tensor<256x12xf32>
  return %9 : tensor<256x12xf32>
}
```

```mlir
// -----// IR Dump After FormDispatchRegions (iree-flow-form-dispatch-regions) //----- //
util.func public @collapse_pack_and_unpack_expand(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub, iree.reflection = {iree.abi.declaration = "sync func @collapse_pack_and_unpack_expand(%input0: tensor<4x3072x64xf32>, %input1: tensor<3072xf32>, %input2: tensor<32x256x8x1xf32>) -> (%output0: tensor<256x12xf32>)"}} {
  %cst = arith.constant 0.000000e+00 : f32
  %0 = hal.tensor.import %arg0 "input0" : !hal.buffer_view -> tensor<4x3072x64xf32>
  %1 = hal.tensor.import %arg1 "input1" : !hal.buffer_view -> tensor<3072xf32>
  %2 = hal.tensor.import %arg2 "input2" : !hal.buffer_view -> tensor<32x256x8x1xf32>
  %3 = tensor.empty() : tensor<3072x64x4xf32>
  %4 = flow.dispatch.region -> (tensor<3072x64x4xf32>) {
    %16 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0, d1)>, affine_map<(d0, d1, d2) -> (d0)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%0, %1 : tensor<4x3072x64xf32>, tensor<3072xf32>) outs(%3 : tensor<3072x64x4xf32>) {
    ^bb0(%in: f32, %in_0: f32, %out: f32):
      %17 = arith.addf %in, %in_0 : f32
      linalg.yield %17 : f32
    } -> tensor<3072x64x4xf32>
    flow.return %16 : tensor<3072x64x4xf32>
  }
  %collapsed = tensor.collapse_shape %4 [[0], [1, 2]] : tensor<3072x64x4xf32> into tensor<3072x256xf32>
  %5 = tensor.empty() : tensor<384x256x8x1xf32>
  %6 = flow.dispatch.region -> (tensor<384x256x8x1xf32>) {
    %pack = tensor.pack %collapsed outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %5 : tensor<3072x256xf32> -> tensor<384x256x8x1xf32>
    flow.return %pack : tensor<384x256x8x1xf32>
  }
  %7 = tensor.empty() : tensor<384x32x8x8xf32>
  %8 = linalg.fill ins(%cst : f32) outs(%7 : tensor<384x32x8x8xf32>) -> tensor<384x32x8x8xf32>
  %9 = flow.dispatch.region -> (tensor<384x32x8x8xf32>) {
    %16 = linalg.mmt4d ins(%6, %2 : tensor<384x256x8x1xf32>, tensor<32x256x8x1xf32>) outs(%8 : tensor<384x32x8x8xf32>) -> tensor<384x32x8x8xf32>
    flow.return %16 : tensor<384x32x8x8xf32>
  }
  %10 = tensor.empty() : tensor<3072x256xf32>
  %11 = flow.dispatch.region -> (tensor<3072x256xf32>) {
    %unpack = tensor.unpack %9 outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 8] into %10 : tensor<384x32x8x8xf32> -> tensor<3072x256xf32>
    flow.return %unpack : tensor<3072x256xf32>
  }
  %expanded = tensor.expand_shape %11 [[0, 1], [2]] : tensor<3072x256xf32> into tensor<12x256x256xf32>
  %12 = tensor.empty() : tensor<256x12xf32>
  %13 = linalg.fill ins(%cst : f32) outs(%12 : tensor<256x12xf32>) -> tensor<256x12xf32>
  %14 = flow.dispatch.region -> (tensor<256x12xf32>) {
    %16 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>, affine_map<(d0, d1, d2) -> (d0, d1)>], iterator_types = ["parallel", "parallel", "reduction"]} ins(%expanded : tensor<12x256x256xf32>) outs(%13 : tensor<256x12xf32>) {
    ^bb0(%in: f32, %out: f32):
      %17 = arith.addf %out, %in : f32
      linalg.yield %17 : f32
    } -> tensor<256x12xf32>
    flow.return %16 : tensor<256x12xf32>
  }
  %15 = hal.tensor.export %14 "output0" : tensor<256x12xf32> -> !hal.buffer_view
  util.return %15 : !hal.buffer_view
}
```

I think potentially we can swap `tensor.collapse_shape` and `tensor.pack` when the original dims are divisible by the inner tile sizes. For example:

```mlir
%collapsed = tensor.collapse_shape %1 [[0], [1, 2]] : tensor<3072x64x4xf32> into tensor<3072x256xf32>
%pack = tensor.pack %collapsed outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %2 : tensor<3072x256xf32> -> tensor<384x256x8x1xf32>
```

can be rewritten into:

```mlir
%pack = tensor.pack %1 outer_dims_perm = [0, 1, 2] inner_dims_pos = [0, 2] inner_tiles = [8, 1] into %2 : tensor<3072x64x4xf32> -> tensor<384x64x4x8x1xf32>
%collapsed = tensor.collapse_shape %pack [[0], [1, 2], [3], [4]] : tensor<384x64x4x8x1xf32> into tensor<384x256x8x1xf32>
```

Then the `tensor.pack` can be fused with its producer. The same transform can also be done on `tensor.unpack`.

It's also worth noting that there are some simpler cases with only unit dims collapsed. For example:

```mlir
#map = affine_map<(d0, d1, d2, d3) -> (d2, d3, d0, d1)>
#map1 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
module {
  func.func @collapse_unit_dim_and_pack(%arg0: tensor<1x256x1x64xf32>, %arg1: tensor<128x64x8x1xf32>) -> tensor<128x32x8x8xf32> {
    %cst = arith.constant 0.000000e+00 : f32
    %0 = tensor.empty() : tensor<1x64x1x256xf32>
    %1 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%arg0 : tensor<1x256x1x64xf32>) outs(%0 : tensor<1x64x1x256xf32>) {
    ^bb0(%in: f32, %out: f32):
      linalg.yield %in : f32
    } -> tensor<1x64x1x256xf32>
    %collapsed = tensor.collapse_shape %1 [[0, 1], [2, 3]] : tensor<1x64x1x256xf32> into tensor<64x256xf32>
    %2 = tensor.empty() : tensor<32x64x8x1xf32>
    %pack = tensor.pack %collapsed outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [8, 1] into %2 : tensor<64x256xf32> -> tensor<32x64x8x1xf32>
    %3 = tensor.empty() : tensor<128x32x8x8xf32>
    %4 = linalg.fill ins(%cst : f32) outs(%3 : tensor<128x32x8x8xf32>) -> tensor<128x32x8x8xf32>
    %5 = linalg.mmt4d ins(%arg1, %pack : tensor<128x64x8x1xf32>, tensor<32x64x8x1xf32>) outs(%4 : tensor<128x32x8x8xf32>) -> tensor<128x32x8x8xf32>
    return %5 : tensor<128x32x8x8xf32>
  }
}
```

@dcaballe fyi

**Comment by pzread** [comment]
Hi @hanhanW, I think you was looking into this issue before. Do you know if some solutions have been proposed before for this case?

**Comment by hanhanW** [comment]
@Max191 I think you hit the same issue about propagating pack across reshape ops. Do you have some patterns in flight?

**Comment by Max191** [comment]
> @Max191 I think you hit the same issue about propagating pack across reshape ops. Do you have some patterns in flight?

I haven't run into this issue before, so I don't have any patterns for it, but it seems like a useful pattern to have. Maybe we can add something here: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Linalg/Transforms/DataLayoutPropagation.cpp

**Comment by pzread** [comment]
Interesting, I didn't know we already have transforms to bubble pack/unpack ops. So looks like we want to do here is to have `BubbleUpPackThroughCollapseShapeOp` and `PushDownUnPackThroughExpandShapeOp`

**Comment by Max191** [comment]
> Interesting, I didn't know we already have transforms to bubble pack/unpack ops. So looks like we want to do here is to have `BubbleUpPackThroughCollapseShapeOp` and `PushDownUnPackThroughExpandShapeOp`

Yes, and there could also be cases where we want to propagate pack through expand_shape or unpack through collapse_shape. Perhaps we can just add `BubbleUpPackThroughReshape` and `PushDownUnPackThroughReshape`, and handle expansion and collapsing.

**Comment by pzread** [comment]
I noticed that `mlir::linalg::populateDataLayoutPropagationPatterns` is only enabled in transform dialect and a specific convolution case. Do we have a plan to enable it globally by default?

**Comment by hanhanW** [comment]
> Do we have a plan to enable it globally by default?

@qedawkins might know the answer. I think it is not ready yet, so they are put at preprocessing layer.

**Comment by hanhanW** [comment]
I just found that I filed https://github.com/openxla/iree/issues/13531 before, but I never get a chance to implement it. Let's use this issue for tracking. @pzread are you going to work on it? I assign the issue to you for now, but feel free to drop if you don't plan to work on it.

**Comment by qedawkins** [comment]
There has been a lot of discussion about layout propagation, but the direction we are moving towards is to avoid materializing pack/unpack in GlobalOptimization/Flow. The DataLayoutPropagation patterns are still useful for path finding in the short term, but they are only opt-in in the preprocessing layer right now, e.g. https://github.com/openxla/iree/blob/e612e91f79f5b95cc73fa3276aefad67860d36a6/compiler/src/iree/compiler/Preprocessing/Common/ConvertConvToChannelsLast.cpp#L683

**Comment by pzread** [comment]
To enable layout propagation in GlobalOptimization, will it be a good idea to add it under a flag? So we can see its impact on the performance first

Also how fast are we moving toward dropping pack/unpack materialization? If it is happening in a few weeks, then I'll probably wait for that; otherwise I'm happy to add these patterns and see how it performs.

**Comment by hanhanW** [comment]
@Max191 is adding propagation prototype under a flag in https://github.com/openxla/iree/pull/16585 

I think we can add the patterns upstream, because they are very useful.

**Comment by Max191** [comment]
> @Max191 is adding propagation prototype under a flag in #16585
> 
> I think we can add the patterns upstream, because they are very useful.

I should note that this propagation does not propagate through reshapes yet. Considering reshapes in the full program analysis makes things pretty difficult, so for now I am not adding any propagation through reshape. I think it could make sense to instead add the upstream patterns as a preprocessing step before the analysis. I actually have another PR coming later that will add a preprocessing step to the analysis pass, so maybe we can try adding it there once everything in flight lands.

**Comment by dcaballe** [comment]
I think adding the patterns to bubble up/down pack/unpack ops through reshapes could be a good starting point that is kind of horizontal to a more generic data layout propagation. The main goal here is not enable fusion of pack/unpack ops with their immediate producer/consumers. We have similar bubble up/down patterns for other ops (`vector.insert/extract` and `vector.bitcast`) with similar transformation enablement purposes. Note that this issue has been a long lasting one: https://github.com/openxla/iree/issues/15242 so having a solution for it in the short term would be awesome now that we understand what is going on!

**Comment by pzread** [comment]
Sent out the propagation patterns to upstream for review: https://github.com/llvm/llvm-project/pull/85297

**Comment by pzread** [comment]
Tested with https://github.com/openxla/iree/pull/16731 and it shows 16% improvement on Falcon7b (x86).

The major regression is on DeepLabV3 and the issue is tracked on https://github.com/openxla/iree/issues/16772.

@hanhanW I saw you start moving materialization away from the global optimization. My current implementation is to swap pack/unpack with reshape at global optimization level. So I'm curious about the timeline on your side and do you think it is worth to land #16731 ?

**Comment by hanhanW** [comment]
> Tested with #16731 and it shows 16% improvement on Falcon7b (x86).

wow, really nice improvements!

> @hanhanW I saw you start moving materialization away from the global optimization. My current implementation is to swap pack/unpack with reshape at global optimization level. So I'm curious about the timeline on your side and do you think it is worth to land #16731 ?

I think it is still worth to push it forward because it shows the optimization opportunities on the early materialization path. We should be able to apply similar optimization in the next evolution of data-tiling. However, I honestly don't have a clear picture about how we propagate encodings through reshapes. My intuition is that we can get such information at stream level or right before converting them to stream. The analysis is much harder because they are formed into dispatches.. The encoding has more information today comparing to the very beginning stage. It has indexing_maps and roles; we might be able to make it work with encodings with some constraints.

Moving the materialization away from global optimization does not mean that the work is a waste. To me, it is more like "we need to consider such chance in the roadmap". If we can't do it in the data-tiling infra, we can move the passes to preprocessing stage. Others still can play with it and demonstrate performance. So yes, the pass is still valuable to me.

**Comment by pzread** [comment]
Merged https://github.com/openxla/iree/pull/16731
